### PR TITLE
Auto-update with remote registry URL

### LINKED ISSUE #12: Monitor tdex-registry URL for new entries

### PR DESCRIPTION
**Pull request by louisinger**
This PR moves the registry logic to `pkg/registry` + set up auto-update logic in case of remote registry

- now `TorProxy` struct has a `Registry` member representing either `ConstantRegistry` (from file or from JSON string) or `RemoteRegistry` from a remote URL.
- Using `StartAutoUpdateRedirects` function, the proxy is now able to set up an auto-updater: refetching the registry from remote URL every 12 hours.

it closes #12 

EDIT: maybe the interval should be configurable via CLI option ?

@tiero please review

### CODE REVIEW

**Review by tiero — CHANGES_REQUESTED** [review]
Yes, let's make the interval configurable

File: pkg/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+type Registry interface {
+	RegistryType() RegistryType
+	GetJSON() ([]byte, error)
+}
+
+// Observe the registry for changes
+// returns json bytes and errors via channels
+func Observe(registry Registry, period time.Duration) (<-chan []byte, <-chan error, func()) {
+	jsonChan := make(chan []byte)
+	errorsChan := make(chan error)
```
Comment (altafan) [inline pkg/registry/registry.go:29]: better to have only one channel of a struct containing the json blob and the error like:
```
make(chan, struct{
   json []byte
   err error
})
```

Give this type a good name :).

File: pkg/torproxy/torproxy.go [hunk 2]
```diff
@@ -61,30 +66,80 @@ func NewTorProxyFromHostAndPort(torHost string, torPort int) (*TorProxy, error)
...
+			select {
+			case json := <-registryJSONchan:
+				err := tp.setRedirectsFromRegistry(json)
+				if err != nil {
+					errorHandler(err)
+				}
+			case err := <-errors:
+				errorHandler(err)
+			}
+		}
```
Comment (altafan) [inline pkg/torproxy/torproxy.go:136]: with the changes to the channel type described above, this should become a simple `for t := range greatChanName { ... }`

**Review by altafan — CHANGES_REQUESTED** [review]
* Better to use `time.Ticker` instead of `time.Sleep` to repeat actions. This [SO](https://stackoverflow.com/questions/53315716/which-one-of-the-two-is-idiomatic-way-time-sleep-or-ticker) explains well the reason.
You can take inspiration from the `Observable/ObservableHandler` design of the daemon's [crawler](https://github.com/tdex-network/tdex-daemon/tree/master/pkg/crawler) to use tickers.
* I think `TorProxy` should expose a `Close()` method that takes care of closing its inner services and channels and whatsoever. This way, the main `startAction` would have only a single deferred function, lowering the risk of forgetting to close something.

**Comment by altafan** [comment]
@louisinger remove [TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA](https://github.com/louisinger/tor-proxy/blob/pull-registry/pkg/torproxy/torproxy.go#L180) from the TLS cypher suites, please.

File: cmd/start.go [hunk 3]
```diff
@@ -15,6 +15,9 @@ import (
 	"github.com/weppos/publicsuffix-go/publicsuffix"
 )
 
+// in case of remote registry, an auto updater is set up
+const autoUpdatePeriod time.Duration = 12 * time.Hour
```
Comment (altafan) [inline cmd/start.go:19]: the update period should be configurable with a flag instead of being hardcoded. Just a reminder.

Comment (tiero) [reply cmd/start.go:19]: cc @louisinger

**Pull request merged**